File: packages/lessmark/src/rules.js
import { BLOCK_ATTRS, CALLOUT_KINDS, DIAGRAM_KINDS, LIST_KINDS, MATH_NOTATIONS, RISK_LEVELS, TASK_STATUSES } from "./grammar.js";

export const HTML_TAG_PATTERN = /<!--|<!doctype\b|<!\[CDATA\[|<\?|<\/?[A-Za-z][A-Za-z0-9:-]*(?:\s[^>]*)?>/i;
export const API_NAME_PATTERN = /^[A-Za-z_][A-Za-z0-9_.-]*$/;
export const CODE_LANG_PATTERN = /^[A-Za-z0-9_.+-]+$/;
export const CONTROL_WHITESPACE_PATTERN = /[\r\n\t]/;
export const MARKDOWN_REFERENCE_DEFINITION_PATTERN = /^\s{0,3}\[[^\]\n]+\]:\s+\S/;
export const MARKDOWN_THEMATIC_BREAK_PATTERN = /^(?:(?: {0,3})(?:[-*_]\s*){3,}|(?: {0,3})=+\s*)$/;
export const MARKDOWN_BLOCKQUOTE_PATTERN = /^\s{0,3}>\s?/;
export const MARKDOWN_LIST_MARKER_PATTERN = /^\s{0,3}(?:[-+*]\s+|\d+[.)]\s+)/;
export const DECISION_ID_PATTERN = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
export const SKILL_NAME_PATTERN = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
export const METADATA_KEY_PATTERN = /^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$/;
export const DEFINITION_TERM_PATTERN = /^(?=.*\S)[^\r\n\t<>]+$/;
export const NAV_SLOTS = new Set(["primary", "footer"]);
export const MAX_INLINE_DEPTH = 128;
export const MAX_LIST_DEPTH = 128;

export function containsRawExpression(text) {
  const source = String(text);
  for (let index = 0; index < source.length; index += 1) {
    const current = source[index];
    const previous = source[index - 1] || "";
    const next = source[index + 1] || "";
    if (current === "$" && next === "{") return true;
    if (current === "{" && previous !== "{" && next !== "{") return true;
    if (current === "}" && previous !== "}" && next !== "}") return true;
  }
  return false;
}

export function isRelativeProjectPath(path) {
  return (
    path.length > 0 &&
    !path.startsWith("/") &&
    !path.startsWith("\\") &&
    !/^[A-Za-z]:[\\/]/.test(path) &&
    !/^[A-Za-z][A-Za-z0-9+.-]*:/.test(path) &&
    !path.split(/[\\/]+/).includes("..")
  );
}

export function isSafeHref(href) {
  const scheme = /^([A-Za-z][A-Za-z0-9+.-]*):/.exec(href);
  if (scheme) return ["http", "https", "mailto"].includes(scheme[1].toLowerCase());
  if (href.startsWith("//")) return false;
  if (href.startsWith("/")) return !href.split(/[\\/]+/).includes("..");
  return isRelativeProjectPath(href);
}

export function isSafeResource(src) {
  const scheme = /^([A-Za-z][A-Za-z0-9+.-]*):/.exec(src);
  if (scheme) return ["http", "https"].includes(scheme[1].toLowerCase());
  return isRelativeProjectPath(src);
}

export function hasBlockAttrSpec(name) {
  return Object.hasOwn(BLOCK_ATTRS, name);
}

export function getBlockAttrErrors(name, attrs) {
  const spec = BLOCK_ATTRS[name];
  if (!spec) {
    return [`Unknown typed block "${name}"`];
  }

  const errors = [];
  for (const key of Object.keys(attrs)) {
    if (!spec.allowed.has(key)) {
      errors.push(`@${name} does not allow attribute "${key}"`);
    }
  }

  for (const key of spec.required) {
    if (!Object.hasOwn(attrs, key) || attrs[key] === "") {
      errors.push(`@${name} requires ${key}`);
    }
  }

  const semanticError = getSemanticAttrError(name, attrs);
  if (semanticError) {
    errors.push(semanticError);
  }

  return errors;
}

export function getBlockBodyErrors(node) {
  const attrs = node.attrs ?? {};
  if (["image", "nav", "page", "separator", "toc"].includes(node.name) && node.text.trim() !== "") {
    return [`@${node.name} must not have a body`];
  }
  if (node.name === "list") {
    return getListBodyErrors(node.text);
  }
  if (node.name === "table") {
    return getTableBodyErrors(attrs.columns ?? "", node.text);
  }
  if (!["code", "example", "math", "diagram"].includes(node.name)) {
    const legacyError = getLegacyMarkdownBodyError(node.text);
    if (legacyError) return [legacyError];
  }
  return [];
}

export function getLegacyMarkdownLineError(line) {
  if (MARKDOWN_REFERENCE_DEFINITION_PATTERN.test(line)) {
    return "Markdown reference definitions are not supported; use @reference or {{ref:label|target}}";
  }
  if (MARKDOWN_THEMATIC_BREAK_PATTERN.test(line)) {
    return "Markdown thematic breaks and setext underlines are not supported; use @separator or # headings";
  }
  if (MARKDOWN_BLOCKQUOTE_PATTERN.test(line)) {
    return "Markdown blockquote markers are not supported in Lessmark source; use @quote or @callout";
  }
  if (MARKDOWN_LIST_MARKER_PATTERN.test(line)) {
    return "Markdown list markers are not supported in Lessmark prose; use @list";
  }
  return null;
}

function getLegacyMarkdownBodyError(text) {
  for (const line of String(text).split("\n")) {
    const error = getLegacyMarkdownLineError(line);
    if (error) return error;
  }
  return null;
}

function getSemanticAttrError(name, attrs) {
  if (name === "task" && typeof attrs.status === "string" && attrs.status && !TASK_STATUSES.has(attrs.status)) {
    return "@task status must be one of: todo, doing, done, blocked";
  }
  if (name === "skill" && typeof attrs.name === "string" && attrs.name) {
    if (!SKILL_NAME_PATTERN.test(attrs.name) || attrs.name.length > 64 || attrs.name.includes("--")) {
      return "@skill name must be 1-64 lowercase letters, numbers, and single hyphens";
    }
  }
  if (name === "skill" && typeof attrs.description === "string" && attrs.description && attrs.description.length > 1024) {
    return "@skill description must be 1-1024 characters";
  }
  if (name === "skill" && typeof attrs.compatibility === "string" && attrs.compatibility && attrs.compatibility.length > 500) {
    return "@skill compatibility must be 1-500 characters";
  }
  if (name === "decision" && typeof attrs.id === "string" && attrs.id && !DECISION_ID_PATTERN.test(attrs.id)) {
    return "@decision id must be a lowercase slug";
  }
  if (name === "file" && typeof attrs.path === "string" && attrs.path && !isRelativeProjectPath(attrs.path)) {
    return "@file path must be a relative project path";
  }
  if (name === "api" && typeof attrs.name === "string" && attrs.name && !API_NAME_PATTERN.test(attrs.name)) {
    return "@api name must be an identifier";
  }
  if (name === "link" && typeof attrs.href === "string" && attrs.href && !isSafeHref(attrs.href)) {
    return "@link href must be http, https, mailto, or a safe relative path";
  }
  if (name === "code" && typeof attrs.lang === "string" && attrs.lang && !CODE_LANG_PATTERN.test(attrs.lang)) {
    return "@code lang must be a compact language identifier";
  }
  if (name === "metadata" && typeof attrs.key === "string" && attrs.key && !METADATA_KEY_PATTERN.test(attrs.key)) {
    return "@metadata key must be a lowercase dotted key";
  }
  if (name === "risk" && typeof attrs.level === "string" && attrs.level && !RISK_LEVELS.has(attrs.level)) {
    return "@risk level must be one of: low, medium, high, critical";
  }
  if (name === "callout" && typeof attrs.kind === "string" && attrs.kind && !CALLOUT_KINDS.has(attrs.kind)) {
    return "@callout kind must be one of: note, tip, warning, caution";
  }
  if (name === "list" && typeof attrs.kind === "string" && attrs.kind && !LIST_KINDS.has(attrs.kind)) {
    return "@list kind must be one of: unordered, ordered";
  }
  if (name === "table" && typeof attrs.columns === "string" && attrs.columns && !isValidTableColumns(attrs.columns)) {
    return "@table columns must be pipe-separated non-empty labels";
  }
  if (name === "image" && typeof attrs.src === "string" && attrs.src && !isSafeResource(attrs.src)) {
    return "@image src must be a safe relative, http, or https URL";
  }
  if (name === "math" && typeof attrs.notation === "string" && attrs.notation && !MATH_NOTATIONS.has(attrs.notation)) {
    return "@math notation must be one of: tex, asciimath";
  }
  if (name === "diagram" && typeof attrs.kind === "string" && attrs.kind && !DIAGRAM_KINDS.has(attrs.kind)) {
    return "@diagram kind must be one of: mermaid, graphviz, plantuml";
  }
  if (name === "nav" && typeof attrs.label === "string" && attrs.label && !DEFINITION_TERM_PATTERN.test(attrs.label)) {
    return "@nav label must be plain single-line text";
  }
  if (name === "nav" && typeof attrs.href === "string" && attrs.href && !isSafeHref(attrs.href)) {
    return "@nav href must be http, https, mailto, or a safe relative path";
  }
  if (name === "nav" && Object.hasOwn(attrs, "slot") && typeof attrs.slot === "string" && !NAV_SLOTS.has(attrs.slot)) {
    return "@nav slot must be primary or footer";
  }
  if (name === "footnote" && typeof attrs.id === "string" && attrs.id && !DECISION_ID_PATTERN.test(attrs.id)) {
    return "@footnote id must be a lowercase slug";
  }
  if (name === "definition" && typeof attrs.term === "string" && attrs.term && !DEFINITION_TERM_PATTERN.test(attrs.term)) {
    return "@definition term must be plain single-line text";
  }
  if (name === "reference" && typeof attrs.target === "string" && attrs.target && !DECISION_ID_PATTERN.test(attrs.target)) {
    return "@reference target must be a lowercase slug";
  }
  if (name === "page" && typeof attrs.output === "string" && attrs.output && !isSafePageOutput(attrs.output)) {
    return "@page output must be a safe relative .html path";
  }
  if (name === "depends-on" && typeof attrs.target === "string" && attrs.target && !DECISION_ID_PATTERN.test(attrs.target)) {
    return "@depends-on target must be a lowercase slug";
  }
  return null;
}

export function isSafePageOutput(output) {
  return isRelativeProjectPath(output) && output.endsWith(".html");
}

export function getLocalAnchorErrors(children) {
  const errors = [];
  const seen = new Set();
  const headingCounts = new Map();
  const targets = new Set();
  for (const node of children) {
    if (node === null || typeof node !== "object") continue;
    let slugs = [];
    if (node.type === "heading") {
      const base = slugifyLocalAnchor(node.text);
      const next = (headingCounts.get(base) || 0) + 1;
      headingCounts.set(base, next);
      slugs = [next === 1 ? base : `${base}-${next}`];
    } else if (node.type === "block" && node.name === "decision") {
      slugs = [node.attrs?.id || ""];
    } else if (node.type === "block" && node.name === "footnote") {
      const id = node.attrs?.id || "";
      slugs = [id, id ? `fn-${id}` : ""];
    }
    for (const slug of slugs) {
      if (!slug) continue;
      if (seen.has(slug)) {
        errors.push(`Duplicate local anchor slug "${slug}"`);
      } else {
        seen.add(slug);
        targets.add(slug);
      }
    }
  }
  for (const node of children) {
    if (node?.type !== "block" || node.name !== "reference") continue;
    const target = node.attrs?.target || "";
    const footnoteTarget = target ? `fn-${target}` : "";
    if (target && !targets.has(target) && !targets.has(footnoteTarget)) {
      errors.push(`Unknown local reference target "${target}"`);
    }
  }
  for (const target of collectInlineLocalTargets(children)) {
    const footnoteTarget = target ? `fn-${target}` : "";
    if (target && DECISION_ID_PATTERN.test(target) && !targets.has(target) && !targets.has(footnoteTarget)) {
      errors.push(`Unknown inline local target "${target}"`);
    }
  }
  return errors;
}

function collectInlineLocalTargets(children) {
  const targets = [];
  for (const node of children) {
    if (node?.type === "heading") {
      targets.push(...inlineTargetsFromText(node.text));
    } else if (node?.type === "block" && !["code", "example", "math", "diagram"].includes(node.name)) {
      targets.push(...inlineTargetsFromText(node.text));
      for (const key of ["label", "cite", "title", "caption", "term"]) {
        if (typeof node.attrs?.[key] === "string") targets.push(...inlineTargetsFromText(node.attrs[key]));
      }
    }
  }
  return targets;
}

function inlineTargetsFromText(text, depth = 0) {
  if (depth > MAX_INLINE_DEPTH) return [];
  const source = String(text);
  const targets = [];
  let index = 0;
  while (index < source.length) {
    const start = source.indexOf("{{", index);
    if (start === -1) break;
    const end = findInlineFunctionEnd(source, start);
    if (end === -1) break;
    const inner = source.slice(start + 2, end);
    const separator = inner.indexOf(":");
    if (separator > 0) {
      const name = inner.slice(0, separator).trim();
      const value = inner.slice(separator + 1);
      if (name === "ref") {
        const delimiter = value.indexOf("|");
        if (delimiter !== -1) targets.push(value.slice(delimiter + 1).trim());
        targets.push(...inlineTargetsFromText(value.slice(0, Math.max(0, delimiter)), depth + 1));
      } else if (name === "footnote") {
        targets.push(value.trim());
      } else if (["strong", "em", "del", "mark", "link"].includes(name)) {
        targets.push(...inlineTargetsFromText(name === "link" ? value.split("|", 1)[0] : value, depth + 1));
      }
    }
    index = end + 2;
  }
  return targets;
}

function findInlineFunctionEnd(source, start) {
  let depth = 1;
  let index = start + 2;
  while (index < source.length) {
    if (source.startsWith("{{", index)) {
      depth += 1;
      index += 2;
      continue;
    }
    if (source.startsWith("}}", index)) {
      depth -= 1;
      if (depth === 0) return index;
      index += 2;
      continue;
    }
    index += 1;
  }
  return -1;
}

function isValidTableColumns(columns) {
  const labels = splitTableRow(columns);
  return labels.length >= 1 && labels.every(Boolean);
}

function getListBodyErrors(text) {
  const lines = String(text).split("\n").filter((line) => line.trim() !== "");
  let previousLevel = 0;
  let seenItem = false;
  for (const line of lines) {
    const match = /^( *)- (.*)$/.exec(line);
    if (!match) {
      return ["@list items must use one explicit '- ' item marker per line"];
    }
    if (match[1].length % 2 !== 0) {
      return ["@list nesting must use two spaces per level"];
    }
    if (match[2].trim() === "") {
      return ["@list items cannot be empty"];
    }
    const level = match[1].length / 2;
    if (!seenItem && level !== 0) {
      return ["@list must start at the top level"];
    }
    if (level > previousLevel + 1) {
      return ["@list nesting cannot skip levels"];
    }
    if (level > MAX_LIST_DEPTH) {
      return ["@list nesting is too deep"];
    }
    if (/\t/.test(line)) {
      return ["@list items must use one explicit '- ' item marker per line"];
    }
    previousLevel = level;
    seenItem = true;
  }
  return [];
}

function getTableBodyErrors(columns, text) {
  const columnCount = splitTableRow(columns).length;
  for (const line of String(text).split("\n").filter((row) => row.trim() !== "")) {
    const cells = splitTableRow(line);
    if (cells.length !== columnCount) return ["@table row cell count must match columns"];
  }
  return [];
}

export function splitTableRow(value) {
  const cells = [];
  let cell = "";
  let escaping = false;
  for (const char of String(value)) {
    if (escaping) {
      if (char !== "|" && char !== "\\") cell += "\\";
      cell += char;
      escaping = false;
      continue;
    }
    if (char === "\\") {
      escaping = true;
      continue;
    }
    if (char === "|") {
      cells.push(cell.trim());
      cell = "";
      continue;
    }
    cell += char;
  }
  if (escaping) cell += "\\";
  cells.push(cell.trim());
  return cells;
}

function slugifyLocalAnchor(text) {
  return String(text)
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "") || "section";
}

File: crates/lessmark/src/rules.rs
use crate::grammar::{
    BlockAttrSpec, BLOCK_ATTR_SPECS, CALLOUT_KINDS, DIAGRAM_KINDS, LIST_KINDS, MATH_NOTATIONS,
    RISK_LEVELS, TASK_STATUSES,
};
use regex::Regex;
use serde_json::Value;
use std::collections::BTreeMap;
use std::sync::OnceLock;

fn html_tag_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r#"</?[A-Za-z][A-Za-z0-9:-]*(?:\s[^>]*)?>"#).unwrap())
}

fn markdown_reference_definition_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"^\s{0,3}\[[^\]\n]+\]:\s+\S").unwrap())
}

fn markdown_thematic_break_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"^(?:(?: {0,3})(?:[-*_]\s*){3,}|(?: {0,3})=+\s*)$").unwrap())
}

fn markdown_blockquote_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"^\s{0,3}>\s?").unwrap())
}

fn api_name_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r#"^[A-Za-z_][A-Za-z0-9_.-]*$"#).unwrap())
}

fn code_lang_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r#"^[A-Za-z0-9_.+-]+$"#).unwrap())
}

fn decision_id_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r#"^[a-z0-9]+(?:-[a-z0-9]+)*$"#).unwrap())
}

fn metadata_key_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r#"^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$"#).unwrap())
}

fn definition_term_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r#"^[^\r\n\t<>]+$"#).unwrap())
}

fn windows_drive_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r#"^[A-Za-z]:[\\/]"#).unwrap())
}

fn uri_scheme_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r#"^([A-Za-z][A-Za-z0-9+.-]*):"#).unwrap())
}

pub fn contains_html_like_tag(text: &str) -> bool {
    html_tag_pattern().is_match(text)
}

pub fn contains_control_whitespace(text: &str) -> bool {
    text.chars().any(|ch| matches!(ch, '\r' | '\n' | '\t'))
}

pub fn get_legacy_markdown_line_error(line: &str) -> Option<&'static str> {
    if markdown_reference_definition_pattern().is_match(line) {
        return Some(
            "Markdown reference definitions are not supported; use @reference or {{ref:label|target}}",
        );
    }
    if markdown_thematic_break_pattern().is_match(line) {
        return Some(
            "Markdown thematic breaks and setext underlines are not supported; use @separator or # headings",
        );
    }
    if markdown_blockquote_pattern().is_match(line) {
        return Some(
            "Markdown blockquote markers are not supported in Lessmark source; use @quote or @callout",
        );
    }
    None
}

pub fn get_block_attr_errors(name: &str, attrs: &BTreeMap<String, String>) -> Vec<String> {
    let Some(spec) = BLOCK_ATTR_SPECS.get(name) else {
        return vec![format!("Unknown typed block \"{}\"", name)];
    };

    let mut errors = Vec::new();
    for key in attrs.keys() {
        if !spec.allowed.contains(key.as_str()) {
            errors.push(format!("@{} does not allow attribute \"{}\"", name, key));
        }
    }
    for key in &spec.required {
        if attrs.get(*key).is_none_or(String::is_empty) {
            errors.push(format!("@{} requires {}", name, key));
        }
    }
    if let Some(error) = semantic_attr_error(name, attrs) {
        errors.push(error);
    }
    errors
}

pub fn get_block_attr_errors_from_value(
    name: &str,
    attrs: &serde_json::Map<String, Value>,
) -> Vec<String> {
    let Some(spec) = BLOCK_ATTR_SPECS.get(name) else {
        return vec![format!("Unknown typed block \"{}\"", name)];
    };

    let mut errors = attr_shape_errors(name, attrs.keys().map(String::as_str), spec);
    for key in &spec.required {
        match attrs.get(*key) {
            None => errors.push(format!("@{} requires {}", name, key)),
            Some(Value::String(value)) if value.is_empty() => {
                errors.push(format!("@{} requires {}", name, key));
            }
            _ => {}
        }
    }
    if let Some(error) = semantic_attr_error_from_value(name, attrs) {
        errors.push(error);
    }
    errors
}

pub fn get_block_body_errors(
    name: &str,
    attrs: &BTreeMap<String, String>,
    text: &str,
) -> Vec<String> {
    if matches!(name, "image" | "nav" | "page" | "separator" | "toc") && !text.trim().is_empty() {
        return vec![format!("@{} must not have a body", name)];
    }
    if name == "list" {
        return get_list_body_errors(text);
    }
    if name == "table" {
        return get_table_body_errors(attrs.get("columns").map(String::as_str).unwrap_or(""), text);
    }
    if !matches!(name, "code" | "example" | "math" | "diagram") {
        if let Some(error) = text.lines().find_map(get_legacy_markdown_line_error) {
            return vec![error.to_string()];
        }
    }
    Vec::new()
}

pub fn get_block_body_errors_from_value(
    name: &str,
    attrs: Option<&serde_json::Map<String, Value>>,
    text: &str,
) -> Vec<String> {
    let string_attrs: BTreeMap<String, String> = attrs
        .into_iter()
        .flat_map(|attrs| attrs.iter())
        .filter_map(|(key, value)| value.as_str().map(|text| (key.clone(), text.to_string())))
        .collect();
    get_block_body_errors(name, &string_attrs, text)
}

fn attr_shape_errors<'a>(
    name: &str,
    keys: impl Iterator<Item = &'a str>,
    spec: &BlockAttrSpec,
) -> Vec<String> {
    keys.filter(|key| !spec.allowed.contains(*key))
        .map(|key| format!("@{} does not allow attribute \"{}\"", name, key))
        .collect()
}

pub fn is_relative_project_path(path: &str) -> bool {
    !path.is_empty()
        && !path.starts_with('/')
        && !path.starts_with('\\')
        && !windows_drive_pattern().is_match(path)
        && !uri_scheme_pattern().is_match(path)
        && !path.split(['/', '\\']).any(|part| part == "..")
}

pub fn is_safe_href(href: &str) -> bool {
    uri_scheme_pattern()
        .captures(href)
        .and_then(|captures| captures.get(1))
        .map_or_else(
            || is_relative_project_path(href),
            |scheme| {
                matches!(
                    scheme.as_str().to_ascii_lowercase().as_str(),
                    "http" | "https" | "mailto"
                )
            },
        )
}

pub fn is_local_slug(value: &str) -> bool {
    decision_id_pattern().is_match(value)
}

pub fn is_safe_resource(src: &str) -> bool {
    uri_scheme_pattern()
        .captures(src)
        .and_then(|captures| captures.get(1))
        .map_or_else(
            || is_relative_project_path(src),
            |scheme| {
                matches!(
                    scheme.as_str().to_ascii_lowercase().as_str(),
                    "http" | "https"
                )
            },
        )
}

fn semantic_attr_error(name: &str, attrs: &BTreeMap<String, String>) -> Option<String> {
    if name == "task" {
        if let Some(status) = attrs.get("status") {
            if !TASK_STATUSES.contains(&status.as_str()) {
                return Some("@task status must be one of: todo, doing, done, blocked".to_string());
            }
        }
    }
    if name == "decision" {
        if let Some(id) = attrs.get("id") {
            if !decision_id_pattern().is_match(id) {
                return Some("@decision id must be a lowercase slug".to_string());
            }
        }
    }
    if name == "file" {
        if let Some(path) = attrs.get("path") {
            if !is_relative_project_path(path) {
                return Some("@file path must be a relative project path".to_string());
            }
        }
    }
    if name == "api" {
        if let Some(api_name) = attrs.get("name") {
            if !api_name_pattern().is_match(api_name) {
                return Some("@api name must be an identifier".to_string());
            }
        }
    }
    if name == "link" {
        if let Some(href) = attrs.get("href") {
            if !is_safe_href(href) {
                return Some(
                    "@link href must be http, https, mailto, or a safe relative path".to_string(),
                );
            }
        }
    }
    if name == "code" {
        if let Some(lang) = attrs.get("lang") {
            if !code_lang_pattern().is_match(lang) {
                return Some("@code lang must be a compact language identifier".to_string());
            }
        }
    }
    if name == "metadata" {
        if let Some(key) = attrs.get("key") {
            if !metadata_key_pattern().is_match(key) {
                return Some("@metadata key must be a lowercase dotted key".to_string());
            }
        }
    }
    if name == "risk" {
        if let Some(level) = attrs.get("level") {
            if !RISK_LEVELS.contains(&level.as_str()) {
                return Some("@risk level must be one of: low, medium, high, critical".to_string());
            }
        }
    }
    if name == "callout" {
        if let Some(kind) = attrs.get("kind") {
            if !CALLOUT_KINDS.contains(&kind.as_str()) {
                return Some(
                    "@callout kind must be one of: note, tip, warning, caution".to_string(),
                );
            }
        }
    }
    if name == "list" {
        if let Some(kind) = attrs.get("kind") {
            if !LIST_KINDS.contains(&kind.as_str()) {
                return Some("@list kind must be one of: unordered, ordered".to_string());
            }
        }
    }
    if name == "table" {
        if let Some(columns) = attrs.get("columns") {
            if !is_valid_table_columns(columns) {
                return Some("@table columns must be pipe-separated non-empty labels".to_string());
            }
        }
    }
    if name == "image" {
        if let Some(src) = attrs.get("src") {
            if !is_safe_resource(src) {
                return Some("@image src must be a safe relative, http, or https URL".to_string());
            }
        }
    }
    if name == "math" {
        if let Some(notation) = attrs.get("notation") {
            if !MATH_NOTATIONS.contains(&notation.as_str()) {
                return Some("@math notation must be one of: tex, asciimath".to_string());
            }
        }
    }
    if name == "diagram" {
        if let Some(kind) = attrs.get("kind") {
            if !DIAGRAM_KINDS.contains(&kind.as_str()) {
                return Some(
                    "@diagram kind must be one of: mermaid, graphviz, plantuml".to_string(),
                );
            }
        }
    }
    if name == "nav" {
        if let Some(label) = attrs.get("label") {
            if label.trim().is_empty() || !definition_term_pattern().is_match(label) {
                return Some("@nav label must be plain single-line text".to_string());
            }
        }
        if let Some(href) = attrs.get("href") {
            if !is_safe_href(href) {
                return Some(
                    "@nav href must be http, https, mailto, or a safe relative path".to_string(),
                );
            }
        }
        if let Some(slot) = attrs.get("slot") {
            if slot != "primary" && slot != "footer" {
                return Some("@nav slot must be primary or footer".to_string());
            }
        }
    }
    if name == "footnote" {
        if let Some(id) = attrs.get("id") {
            if !decision_id_pattern().is_match(id) {
                return Some("@footnote id must be a lowercase slug".to_string());
            }
        }
    }
    if name == "definition" {
        if let Some(term) = attrs.get("term") {
            if term.trim().is_empty() || !definition_term_pattern().is_match(term) {
                return Some("@definition term must be plain single-line text".to_string());
            }
        }
    }
    if name == "reference" {
        if let Some(target) = attrs.get("target") {
            if !decision_id_pattern().is_match(target) {
                return Some("@reference target must be a lowercase slug".to_string());
            }
        }
    }
    if name == "page" {
        if let Some(output) = attrs.get("output") {
            if !is_safe_page_output(output) {
                return Some("@page output must be a safe relative .html path".to_string());
            }
        }
    }
    if name == "depends-on" {
        if let Some(target) = attrs.get("target") {
            if !decision_id_pattern().is_match(target) {
                return Some("@depends-on target must be a lowercase slug".to_string());
            }
        }
    }
    None
}

fn is_valid_table_columns(columns: &str) -> bool {
    let labels = columns.split('|').map(str::trim).collect::<Vec<_>>();
    !labels.is_empty() && labels.iter().all(|column| !column.is_empty())
}

fn get_list_body_errors(text: &str) -> Vec<String> {
    let mut previous_level = 0usize;
    let mut seen_item = false;
    for line in text.lines().filter(|line| !line.trim().is_empty()) {
        if line.contains('\t') {
            return vec!["@list items must use one explicit '- ' item marker per line".to_string()];
        }
        let Some(marker_index) = line.find("- ") else {
            return vec!["@list items must use one explicit '- ' item marker per line".to_string()];
        };
        if !line[..marker_index].chars().all(|ch| ch == ' ') || marker_index % 2 != 0 {
            return vec!["@list nesting must use two spaces per level".to_string()];
        }
        if line[marker_index + 2..].trim().is_empty() {
            return vec!["@list items cannot be empty".to_string()];
        }
        let level = marker_index / 2;
        if !seen_item && level != 0 {
            return vec!["@list must start at the top level".to_string()];
        }
        if level > previous_level + 1 {
            return vec!["@list nesting cannot skip levels".to_string()];
        }
        previous_level = level;
        seen_item = true;
    }
    Vec::new()
}

fn get_table_body_errors(columns: &str, text: &str) -> Vec<String> {
    let column_count = split_table_row(columns).len();
    for line in text.lines().filter(|line| !line.trim().is_empty()) {
        let cells = split_table_row(line);
        if cells.iter().any(String::is_empty) {
            return vec!["@table cells cannot be empty".to_string()];
        }
        if cells.len() != column_count {
            return vec!["@table row cell count must match columns".to_string()];
        }
    }
    Vec::new()
}

pub fn split_table_row(value: &str) -> Vec<String> {
    let mut cells = Vec::new();
    let mut cell = String::new();
    let mut escaping = false;
    for ch in value.chars() {
        if escaping {
            if ch != '|' && ch != '\\' {
                cell.push('\\');
            }
            cell.push(ch);
            escaping = false;
            continue;
        }
        if ch == '\\' {
            escaping = true;
            continue;
        }
        if ch == '|' {
            cells.push(cell.trim().to_string());
            cell.clear();
            continue;
        }
        cell.push(ch);
    }
    if escaping {
        cell.push('\\');
    }
    cells.push(cell.trim().to_string());
    cells
}

fn is_safe_page_output(output: &str) -> bool {
    is_relative_project_path(output) && output.ends_with(".html")
}

fn semantic_attr_error_from_value(
    name: &str,
    attrs: &serde_json::Map<String, Value>,
) -> Option<String> {
    let string_attrs: BTreeMap<String, String> = attrs
        .iter()
        .filter_map(|(key, value)| value.as_str().map(|text| (key.clone(), text.to_string())))
        .collect();
    semantic_attr_error(name, &string_attrs)
}
